### PR TITLE
12963 google backend

### PR DESCRIPTION
**Pull request by javitonino**
#12963

**Acceptance**

Test the **organization** login and signup pages. `org.carto.com/login` and `org.carto.com/signup`

| Domainful + central      | Login                  | Signup                 | Signup (taken username) |
| ------------- |:----------------------:|:----------------------:|:-----------------------:|
| Google        | <ul><li>[x] </li></ul> | <ul><li>[x] </li></ul> | <ul><li>[x] </li></ul>  |
| GitHub        | <ul><li>[x] </li></ul> | <ul><li>[x] </li></ul> | <ul><li>[x] </li></ul>  |

| Subdomainless - no central  | Login                  | Signup                 | Signup (taken username) |
| ------------- |:----------------------:|:----------------------:|:-----------------------:|
| Google        | <ul><li>[x] </li></ul> | <ul><li>[x] </li></ul> | <ul><li>[x] </li></ul>  |
| GitHub        | <ul><li>[x] </li></ul> | <ul><li>[x] </li></ul> | <ul><li>[x] </li></ul>  |

### CODE REVIEW

File: lib/oauth/google/api.rb [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+      def get_user_data
+        response = Typhoeus::Request.new(
+          "https://www.googleapis.com/plus/v1/people/me?access_token=#{@access_token}",
+          method: 'GET',
+          ssl_verifypeer: true,
+          timeout: 600
+        ).run
+
+        raise 'Invalid response code' unless response.code == 200
+          JSON.parse(response.body)
```
[the remark this answers is not in the document]
Comment (amartinj) [reply lib/oauth/google/api.rb]: Indentation

File: lib/oauth/github/api.rb [hunk 2]
```diff
@@ -75,30 +83,38 @@ def user_emails
...
-          nil
-        end
-      end
-      private_class_method :exchange_code_for_token
-
       def authenticated_request(method, url, body: nil)
         self.class.request(method, url, body: body, headers: { 'Authorization' => "token #{@access_token}" })
       end
+
+      def self.request(method, url, body: nil, headers: {})
```
[the remark this answers is not in the document]
Comment (amartinj) [reply lib/oauth/github/api.rb]: Maybe you should move this method out of private block

[the remark this answers is not in the document]
Comment (amartinj) [reply lib/oauth/github/api.rb]: Same here

[the remark this answers is not in the document]
Comment (javitonino) [reply lib/oauth/github/api.rb]: I'll just add `private_class_method` and keep it under private. No need to go around exposing things :)

**Comment by javitonino** [comment]
@elenatorro found a 500: https://rollbar.com/carto/CartoDB/items/25020/ Thanks!

Back to dev (I hate rails autoloading 😢)

**Comment by javitonino** [comment]
@amartinj We had some issues with namespacin and the `oauth` gem, so I just moved everything to the `Carto` namespace. Can you take a quick look?

**Comment by elenatorro** [comment]
Question: when login in with google / github, if the user doesn't exist, it should redirect her to the sign up page, right?

**Comment by elenatorro** [comment]
Tested on ded14 and not working. I'm writing here the steps I've followed just in case I did something wrong:

Server: ded14

Steps: 
- Create organization and organization user (owner)
- New incognito window
- user + password log in (owner)
- Add `gmail.com` to whitelisted emails (testing it with my email)
- Log out
- Go to organization signup page 
- Click on github button
- Authorize github (two factor auth etc...)
- Redirected to login (organization login page)
- Click on github again
- Redirected back to login (not working)
- Go to organization signup page 
- Click on google button
- Authorize google (two factor auth etc...)
- Redirected to login (organization login page)
- Click on google again
- Redirected back to login (not working)

Note: no errors shown in the console.

Any clue? :pray:

**Comment by javitonino** [comment]
Took a look with @elenatorro and it seems the configuration of oauth is alright, so it seems like a proper bug. Back to dev!

**Comment by javitonino** [comment]
Actually, there were some config changes needed as well. I pushed a code fix and some config changes.

**Comment by amartinj** [comment]
Moving to acceptance again :)

**Comment by amartinj** [comment]
Moving to acceptance

**Comment by elenatorro** [comment]
Tested on ded14: GitHub works fine, but with Google button it's still not working.

**Comment by javitonino** [comment]
There is some misconfiguration in staging. I'll review it.

**Comment by javitonino** [comment]
Ok, sorry for all the configuration mess. Some more info in [here](https://github.com/CartoDB/companystuff/issues/1535). Anyway, it should be testable now. Sorry for all the comings and goings 🙏

**Comment by elenatorro** [comment]
Acceptance steps work fine! But I have a little question: when the user can't sign up / log in through oauth because the email is wrong, it redirects her to this page: 

captura de pantalla 2017-11-02 a las 11 14 49

But the user can't change the email in this field (the input is `readonly`). Is it a bug, or is it the expected behaviour?  Thanks!

**Comment by noguerol** [comment]
Shouldn't that page be the normal signup and not the orgs login page?

**Pull request merged**